### PR TITLE
Using Unicode & `char`

### PR DESCRIPTION
**Pull request by ocornoc**
* Added a .gitignore
* Changed ugly `&**` patterns into `.as_mut()` where possible
* Fixed a few typos in error messages
* Changed use of `u8` into `char` to support Unicode
* Removed restriction that variables must be lowercase ASCII

### CODE REVIEW

**Comment by ocornoc** [comment]
I wish I would've know you'd be posting an update just now! I'm happy to redo this for the new commit if you'd like.

**Comment by christianpoveda** [comment]
Oh hi! sorry i didn't realized of the PR ^^".

I like most of the changes. However, I'm conflicted about the `u8` -> `char` transition. Not accepting unicode is certainly a limitation but I don't know if its actually a big deal. I'll add some comments about it

File: src/parser.rs [hunk 1]
```diff
@@ -17,15 +17,15 @@ pub enum NamedTerm {
...
+            NamedTerm::Var(var) => write!(f, "{}", var),
+            NamedTerm::Abs(var, term) => write!(f, "(λ{}. {})", var, term),
             NamedTerm::App(t1, t2) => write!(f, "({} {})", t1, t2),
         }
     }
 }
 
 pub fn parse(input: &str) -> Option<NamedTerm> {
-    let tokens = Tokenizer::new(input.as_bytes());
+    let tokens = Tokenizer::new(input.chars().collect());
```
Comment (christianpoveda) [inline src/parser.rs:28]: Here we are basically cloning the whole string right?

Edit: The good part about restricting to just ascii is that we can easily point where the syntax errors are by just using `Tokenizer::loc`

**Comment by ocornoc** [comment]
The only reason why I added that was so I could use the lambda character (I'm forking `pierce` to support the mu operator in lambda-mu calculus, so being able to distinguish the two that way is nice).
In a few/many hours, I'll redo the PR as to avoid merge conflicts and I'll skip the unicode support.
That work?

File: src/parser/tokenizer.rs [hunk 2]
```diff
@@ -33,27 +33,26 @@ impl TokenKind {
...
-            b'.' => Dot,
-            b' ' => Space,
-            byte if byte.is_ascii_lowercase() => Char(byte),
-            _ => return Err(ParseError::new(self.loc, ParseErrorKind::InvalidByte(byte))),
+            '(' => LBracket,
+            ')' => RBracket,
+            '\\' | 'λ' => Lambda,
+            '.' => Dot,
+            ' ' => Space,
+            byte => Char(byte),
```
Comment (christianpoveda) [inline src/parser/tokenizer.rs:55]: I think variable names should be at least alphabetic. Otherwise people might write stuff like `(\[. (\]. ([ ])))`

**Comment by christianpoveda** [comment]
I think the ascii restriction should apply only for variables, using the lambda character alongside `\` is ok for me tho. 

I've never heard about the lambda-mu calculus. But after skimming wikipedia I wonder if it's possible to introduce the new expressions as derived forms like `let` bindings.

**Comment by ocornoc** [comment]
Because you're using (or were, i haven't checked the new commit) a byte-by-byte parser (rather than `char`), allowing lambda will require some state. You can just use a `char` parser and still only allow ASCII alphabetic lowercase, though.

I had planned on using either `/` or `|` for the mu operator (alongside `μ`), but now that you have `let`, a simple `mu` might work (maybe `lam` as an alias for lambda too? It's common in other languages and helps with symmetry).

**Comment by ocornoc** [comment]
As a bonus to support Unicode `λ` (and `μ`) is that _iirc_ the [non-error output](https://github.com/christianpoveda/pierce/blob/e83c9dabc0f198f9956acbb8d1c7186f8948a4f9/src/eval.rs#L14) from your program is itself executable by your program.

File: src/parser.rs (hunk 1 above)
Comment (ocornoc) [reply src/parser.rs:28]: Assuming you use `char`s throughout, I believe `loc` would still be indexing by the `n`th `char`, so it'd still work out.

Comment (ocornoc) [reply src/parser.rs:28]: Also, I do believe we're cloning there. That being said, the input source is typically tiny and, without glancing at the code first, I think we could pass by move there if `input` is instead a `String`.

File: src/parser/tokenizer.rs (hunk 2 above)
Comment (ocornoc) [reply src/parser/tokenizer.rs:55]: That makes sense. We can use something like [`char::is_alphabetic`](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphabetic).

File: src/parser.rs (hunk 1 above)
Comment (christianpoveda) [reply src/parser.rs:28]: > Assuming you use `char`s throughout, I believe `loc` would still be indexing by the `n`th `char`, so it'd still work out.

Is it there any guarantee that every `char` uses exactly one cell in the terminal?

Comment (christianpoveda) [reply src/parser.rs:28]: > Also, I do believe we're cloning there. That being said, the input source is typically tiny and, without glancing at the code first, I think we could pass by move there if `input` is instead a `String`.

Maybe we can get some middle ground using `Iterator<Item=char>` instead?

Comment (ocornoc) [reply src/parser.rs:28]: That would be an issue with the fonts people use in their terminals, but to the best of my knowledge, popular monospace fonts (as used in terminals, too) would have the same space for anything people would want to put in. While you can go out of your way to find a couple counterexamples with esoteric characters, frankly, garbage in, garbage out.

File: src/parser/tokenizer.rs (hunk 2 above)
Comment (christianpoveda) [reply src/parser/tokenizer.rs:55]: Give me a couple hours to make my mind about this. I had planned to change the `TokenKind::Word` variant to be just an slice of the original input but that might end up causing more trouble than is worth.

File: src/parser.rs (hunk 1 above)
Comment (ocornoc) [reply src/parser.rs:28]: If you want to go completely generic, using a generic over `I: IntoIterator<char>` would be the best approach.

**Comment by christianpoveda** [comment]
> Because you're using (or were, i haven't checked the new commit) a byte-by-byte parser (rather than `char`), allowing lambda will require some state. You can just use a `char` parser and still only allow ASCII alphabetic lowercase, though.
> 
> I had planned on using either `/` or `|` for the mu operator (alongside `μ`), but now that you have `let`, a simple `mu` might work (maybe `lam` as an alias for lambda too? It's common in other languages and helps with symmetry).

I think using `lam` is fine, or `lambda`.

**Comment by christianpoveda** [comment]
> As a bonus to supporting Unicode `λ` (and `μ`) is that _iirc_ the [non-error output](https://github.com/christianpoveda/pierce/blob/e83c9dabc0f198f9956acbb8d1c7186f8948a4f9/src/eval.rs#L14) from your program is itself executable by your program.

yes that would be awesome.

File: src/parser/tokenizer.rs (hunk 2 above)
Comment (ocornoc) [reply src/parser/tokenizer.rs:55]: I won't be able to start working on this for a while anyways. If I don't hear back from you by the time I'm done with what I'm working on at the moment, I'll create a version just using `char` so we can compare it on the new version.

**Comment by ocornoc** [comment]
I'll put the stuff not having to do with Unicode/`char` into a separate PR sometime today so we can get those out of the way.

**Comment by christianpoveda** [comment]
Sure thanks!

I just added a couple of helper methods to handle words and letters when doing parsing.

**Comment by ocornoc** [comment]
That should make it easier then. :)

**Comment by ocornoc** [comment]
Gonna make this into two separate PRs

**Pull request closed without merging**